subsequences

malarbol opened the pull request.

malarbol commented:
Hello again.
I've been playing around with metric structures and things like that. On the way I ended up needing/wanting these few new properties on sequences. They result interesting to prove things like ("asymptotical equality of sequences preserves limits", or "a sequence has a limit if and only of all its subsequence have this limit"). I think maybe these concepts could also prove themselves interesting in other contexts (e.g. polynomials on ring as asymptotically vanishing sequences, etc.).

malarbol commented:
Hello again.
I've been playing around with metric structures and things like that. On the way I ended up needing/wanting these few new properties on sequences. They result interesting to prove things like ("asymptotical equality of sequences preserves limits", or "a sequence has a limit if and only of all its subsequence have this limit"). I think maybe these concepts could also prove themselves interesting in other contexts (e.g. polynomials on ring as asymptotically vanishing sequences, etc.).

fredrik-bakke commented:
Hey @malarbol, I'm just having a quick look at your changes for now, but why not have separate files for increasing and decreasing sequences? I would similarly expect us to have separate files for order-preserving and order-reversing maps

malarbol commented:
> Hey @malarbol, I'm just having a quick look at your changes for now, but why not have separate files for increasing and decreasing sequences? I would similarly expect us to have separate files for order-preserving and order-reversing maps

Hey @fredrik-bakke, thanks for the feedback. I'm sorry, this PR got a bit bigger than anticipated, (again 😅) and I still have a lot of cleanup to do. I'll do my best to address your concern; we may still need a module importing both of them, for properties like "a sequence is constant iff it is both increasing and decreasing".

fredrik-bakke commented:
That's okay. The property you mentioned should go in a file abour constant sequences :)

malarbol commented:
Hey again @fredrik-bakke.
I refactored a few concepts and cleaned things up a bit. I also updated the title/description of the PR to reflect better its content.
If you prefer, maybe we could split this PR into two, with "new concepts" on one hand, and the "illustrative modules" on the other. Some of these modules could also be more interesting, like maybe proving that "decreasing sequences of natural numbers are asymptotically constant" is equivalent to the LPO, or study behavior of bounded increasing sequences of natural numbers but I'm not sure how to handle these right now.

I already have a few follow-up ideas that motivated this PR:
- metric structures (with sequential limits, etc.)
- series, polynomials, convolution products, etc. (using asymptotically vanishing sequence in monoids, rings, etc.)

fredrik-bakke commented:
Hey Malarbol! I'm back. Sorry for the terribly long wait; I'll try to review your PR in one of the coming days :)

fredrik-bakke commented on src/elementary-number-theory/decreasing-sequences-natural-numbers.lagda.md:
```diff
@@ -0,0 +1,397 @@
+
+## Idea
+
+A sequence of natural numbers is **decreasing** if it reverses
```
```suggestion
A [sequence](foundation.sequences.md) of [natural numbers](elementary-number-theory.natural-numbers.md) is {{#concept "decreasing" Disambiguation="sequence of natural numbers" Agda=is-decreasing-sequence-ℕ}} if it reverses
```

fredrik-bakke commented on src/foundation/asymptotical-dependent-sequences.lagda.md:
```diff
@@ -0,0 +1,157 @@
+
+## Idea
+
+A dependent sequence `A : ℕ → UU l` is **asymptotical** if `A n` is pointed for
```
If you have a dependent sequence `a` of `A : ℕ → UU l` then isn't every `A n` inhabited by `a n`? The sequence `A : ℕ → UU l` isn't dependent.

fredrik-bakke commented on src/foundation/asymptotical-dependent-sequences.lagda.md:
```diff
@@ -0,0 +1,157 @@
+
+## Idea
+
+A dependent sequence `A : ℕ → UU l` is **asymptotical** if `A n` is pointed for
+sufficiently large natural numbers `n`.
```
Is "asymptotical sequence" standard terminology? I can't say I have heard it before. Would better terminology be "asymptotically pointed sequence of types"?

fredrik-bakke commented on src/foundation/asymptotical-value-sequences.lagda.md:
```diff
@@ -0,0 +1,123 @@
+
+## Idea
+
+A sequence `u` in a type `A` has an **asymptotical value** `x : A` if `x ＝ u n`
+for sufficiently large natural numbers `n`.
```
Please add links to the concepts and use the `{{#concept ...}}` macro. this applies to all of your new files.

fredrik-bakke commented on src/foundation/sequences.lagda.md:
````diff
@@ -28,6 +31,17 @@ sequence : {l : Level} → UU l → UU l
 sequence A = dependent-sequence (λ _ → A)
 ```
 
+### Values of a sequence
+
+```agda
+module _
+  {l : Level} {A : UU l}
+  where
+
+  is-value-sequence : A → sequence A → ℕ → UU l
+  is-value-sequence x u n = x ＝ u n
+```
````
I'm not sure this definition is aiding readability, perhaps it is better to just write out `x ＝ u n`?

fredrik-bakke commented on src/foundation/asymptotical-dependent-sequences.lagda.md:
```diff
@@ -0,0 +1,157 @@
+  is-modulus-dependent-sequence : ℕ → UU l
+  is-modulus-dependent-sequence N = (n : ℕ) → leq-ℕ N n → A n
+
+  asymptotically : UU l
+  asymptotically = Σ ℕ is-modulus-dependent-sequence
```
While I can appreciate that this definition makes some of your other code readable and intuitive once you know what it means, I can't say that this name should be reserved for this definition. Two questions I have immediately are
- why would you reserve asymptotics for sequences only over natural numbers
- shouldn't "asymptotically" be a property? If I say something holds asymptotically, would you expect it to matter _how_ it holds asymptotically?

fredrik-bakke commented on src/foundation/asymptotical-value-sequences.lagda.md:
```diff
@@ -0,0 +1,123 @@
+
+## Definitions
+
+### Asymptotical values of sequences
```
```suggestion
### The Asymptotic value of an asymptotic sequence
```

fredrik-bakke commented on src/foundation/asymptotical-value-sequences.lagda.md:
```diff
@@ -0,0 +1,123 @@
+  {l : Level} {A : UU l} (x : A) (u : sequence A)
+  where
+
+  is-∞-value-sequence : UU l
+  is-∞-value-sequence = asymptotically (is-value-sequence x u)
```
We avoid using unicode characters like `∞` in entry names except for in select rare cases like the namespace for the natural numbers `ℕ`. A better name for this entry would be `takes-value-asymptotically` or `value-at-infinity-equals`

fredrik-bakke commented on src/foundation/asymptotical-value-sequences.lagda.md:
```diff
@@ -0,0 +1,123 @@
+# Asymptotical value of a sequence
```
Isn't "asymptotic" the correct conjugation of the word? Please check and if so review your contribution to reflect the correct conjugation

fredrik-bakke commented on src/elementary-number-theory/strictly-increasing-sequences-natural-numbers.lagda.md:
```diff
@@ -0,0 +1,325 @@
+  (f : sequence ℕ)
+  where
+
+  is-strict-increasing-prop-sequence-ℕ : Prop lzero
+  is-strict-increasing-prop-sequence-ℕ =
+    Π-Prop ℕ
+      ( λ i →
+        Π-Prop ℕ
+          ( λ j → hom-Prop (le-ℕ-Prop i j) (le-ℕ-Prop (f i) (f j))))
```
It's more appropriate to define strictly increasing sequences as order homomorphisms of the strict ordering on the natural numbers.

fredrik-bakke commented on src/elementary-number-theory/strictly-increasing-sequences-natural-numbers.lagda.md:
```diff
@@ -0,0 +1,325 @@
+  (f : sequence ℕ)
+  where
+
+  is-strict-increasing-prop-sequence-ℕ : Prop lzero
+  is-strict-increasing-prop-sequence-ℕ =
```
This should be named `is-strictly-increasing-...`

fredrik-bakke commented on src/elementary-number-theory/strictly-decreasing-sequences-natural-numbers.lagda.md:
```diff
@@ -0,0 +1,139 @@
+  (f : sequence ℕ)
+  where
+
+  is-strict-decreasing-sequence-ℕ : UU lzero
+  is-strict-decreasing-sequence-ℕ =
+    (i j : ℕ) → le-ℕ i j → le-ℕ (f j) (f i)
```
this should be named `is-strictly-decreasing-...`

fredrik-bakke commented:
Also, feel free to reopen at any point @malarbol, your contributions are most welcome!

malarbol commented:
> I'm closing this PR as it seems to have been abandoned. I'll note that the work that is done here seems highly relevant to #1211 perhaps some of this work can be salvaged @EgbertRijke?

Hey @fredrik-bakke! I've been wanted to get in touch for some months now but I couldn't find the time / courage / motivation, etc. I'm quite sorry. I've been busy with work, and other side-projects and I haven't written any agda since my PR on metric spaces. But I do miss it so I might get back to it sometimes soon.

I should have closed this PR for month. I opened it to share some of my thoughts around convergence / asymptotical behavior but that was never really ready to be merged. Obviously, feel free to salvage anything you find interesting. Or we can talk about it and restart it from scratch with a better setting.

> Also, feel free to reopen at any point @malarbol, your contributions are most welcome!

Or I'll reopen it if I manage to clean it up. We'll see. 

I see there's been a lot of contributions in the `real-numbers` module, that's really nice! I'll have to read it carefully but it's already giving me some ideas. Is anybody working towards Cauchy-completeness of the real numbers? It seems that the recent additions on the density of Q in R was the missing piece to link real numbers and rational Cauchy approximations. 

Take care.
I look forward to contribute again.

PS: is there any feedback on the `metric-spaces` module? People complaining about the name scheme or coming up with better ideas? I feel like I told you I would think about it and fix the names, and never really did it. Sorry.
